airflow/ti_deps/deps/mapped_task_upstream_dep.py

```diff
@@ -0,0 +1,92 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from __future__ import annotations
+
+from collections.abc import Iterator
+from typing import TYPE_CHECKING
+
+from airflow.ti_deps.deps.base_ti_dep import BaseTIDep
+from airflow.utils.state import State, TaskInstanceState
+
+if TYPE_CHECKING:
+    from sqlalchemy.orm import Session
+
+    from airflow.models.taskinstance import TaskInstance
+    from airflow.ti_deps.dep_context import DepContext
+    from airflow.ti_deps.deps.base_ti_dep import TIDepStatus
+
+
+class MappedTaskUpstreamDep(BaseTIDep):
+    """
+    Determines if a mapped task's upstream tasks that provide XComs used by this task for task mapping are in
+    a state that allows a given task instance to run.
+    """
+
+    NAME = "Mapped dependencies have succeeded"
+    IGNORABLE = True
+    IS_TASK_DEP = True
+
+    def _get_dep_statuses(
+        self,
+        ti: TaskInstance,
+        session: Session,
+        dep_context: DepContext,
+    ) -> Iterator[TIDepStatus]:
+        from airflow.models.mappedoperator import MappedOperator
+
+        if isinstance(ti.task, MappedOperator):
+            mapped_dependencies = ti.task.iter_mapped_dependencies()
+        elif (task_group := ti.task.get_closest_mapped_task_group()) is not None:
```

Comment:
I’m not sure about this (but maybe I’m wrong). If a task is in nested mapped task groups, I _think_ we need to collect dependencies from all those groups, instead of just the innermost one. Maybe this should be something like

```python
itertools.chain.from_iterable(
    tg.iter_mapped_dependencies() for tg in ti.task.iter_mapped_task_groups()
)
```

Same goes with the MappedOperator case, we need to consider task groups as well as the operator itself.

It may be worthwhile to add a function on AbstractOperator to extract this logic. (Not sure either, this can be explored later.)

Reply:
I was thinking about this, but realized that nested mapped task groups and mapped operators within mapped task groups are currently not supported? I now added a couple of unit tests that verify that this is not the case. This way, if they end up being supported in the future there will be failing tests indicating that some work needs to be done.

Reply:
I feel it’s not _that_ difficult to just support the general case here, but deferring the work with an explcit check sounds good enough to me.